### PR TITLE
IMP: transformers SingleLanePerSample... to Casava

### LINKED ISSUE #209: Transform SLPSSEFDF -> CasavaOneEightSingleLanePerSampleDirFmt
**Improvement Description**
Add a transformer from `SingleLanePerSampleSingleEndFastqDirFmt`  to `CasavaOneEightSingleLanePerSampleDirFmt`.

This is convenient for exporting raw reads without the metadata/manifest files laying around.

### PR DESCRIPTION
**Pull request by thermokarst**
Fixes #209

Related to https://github.com/qiime2/q2-quality-control/pull/52/

---

Missing unit tests

### CODE REVIEW

**Comment by thermokarst** [comment]
Relevant discussion here: https://github.com/qiime2/q2-quality-control/pull/52/#issuecomment-634320573

I'm not sure yet about the transformer here

https://github.com/qiime2/q2-types/pull/245/files#diff-3fd5d1462a2b3d80c0d3c98949f10949R94-R102

The problem is that it ends up producing a useless basepath, which ultimately needs to be rewritten in the format property. Perhaps it would be better to eliminate that transformer altogether and stick it all in the property, I was just trying to recycle as much of the transformers functionality as I possibly could. Open to ideas.

**Comment by thermokarst** [comment]
> I'm not sure yet about the transformer here

The more I think about this, the more I want to refactor as internal utils used by the fmt property && the transformers, and dropping this transformer.

**Comment by thermokarst** [comment]
Okay, the diff is pretty gross because there is a bit of code churn, but the refactor looks like this: first I moved all the utils out of `_transformer.py` and into a new shared `_util.py`. The new util module is used by the transformers module && the formats module, so I had to refactor the utils so that they injected their format dependencies at runtime (I am open to other suggestions, just let me know). Then I created some new dynamic utils via `functool`, to cut down on the boilerplate. Lastly, I refactored the dynamic manifest property on the format to use the shared utils.

**Comment by thermokarst** [comment]
BTW - this is still missing unit tests, but I can add those as part of my response-to-review.

File: q2_types/per_sample_sequences/tests/test_format.py [hunk 1]
```diff
@@ -328,6 +329,17 @@ def test_casava_one_eight_slanepsample_dir_fmt_validate_positive(self):
...
 
+    def test_casava_one_eight_slanepsample_dir_fmt_manifest_property(self):
+        filepath = self.get_data_path('Human-Kneecap_S1_L001_R1_001.fastq.gz')
+        shutil.copy(filepath, self.temp_dir.name)
+
+        format = CasavaOneEightSingleLanePerSampleDirFmt(
+            self.temp_dir.name, mode='r')
+
+        format.validate()
+        self.assertTrue(True)
```
[the remark this answers is not in the document]
Comment (thermokarst) [reply q2_types/per_sample_sequences/tests/test_format.py:340]: Yeah, its a checkpoint - I can either add a comment explaining that, or refactor to

```python
self.assertEqual(format.validate(), None)
```

Your call!

[the remark this answers is not in the document]
Comment (nbokulich) [reply q2_types/per_sample_sequences/tests/test_format.py:340]: oh got it! I see now. I will just merge as-is. Thanks!

**Pull request merged**